### ISSUE TITLE
Determining User Requirements

### ISSUE THREAD

**Issue by bahricanyesil** [bounswe/bounswe2022group2#16]
## Description

As we have learned in the lesson, we should firstly determine the requirements of the project. It's the initial step before starting to work further. There are different types of requirements and we decided to collect them under distinct titles. Besides, we also decided to assign different people for each type of requirements and a person who is not assigned to that title's issue to review it. We all will read each requirements issue and we are open to comment further, add new requirements to those titles even we are not assigned to them.

Our base titles for the requirements for now are as follows:
- Functional Requirements
    * User Requirements
    * System Requirements
- Non-Functional Requirements

This structure may be enhanced/enlarged and new base titles can be added in the future. If you can find, you can write them as a  comment to this issue.

According to our decision in meeting 2, responsible people for the **user requirements** are:
- [x] @ahmet633 
- [x] @xltvy - [Profile page requirements](https://github.com/bounswe/bounswe2022group2/issues/19)
- [x] @bahricanyesil - [Authentication related requirements](https://github.com/bounswe/bounswe2022group2/issues/17)
- [x] @mbatuhancelik  [User Interaction requirements](https://github.com/bounswe/bounswe2022group2/issues/23)

These people will create separate issue for the subtitle s/he work and will link those issues under this main issue with a summary comment. We will manage all user requirement related things here with the help of hyperlinks, comments and so on.

There is a one reviewer for this issue and she will review what the assignees found, what others are commented at the end and will make the last sentences before closing the issue to be sure that everything works well.

**Deadline: 13.03.2022 - Sunday - 20:00**
**Reviewer: @ezgy** - Deadline: 14.03.2022 - Monday - 23:59

**Comment by bahricanyesil** [comment]
**INFO:**
I will be detailing the items for the **authentication process** as a subtitle of user requirements.
I will create a new issue for that subtitle and I will link it here.

**Comment by bahricanyesil** [comment]
I will create a separate wiki page for the requirements and we can collect all requirements under that wiki page after the reviews.

**Comment by xltvy** [comment]
**INFO:**
I will identify the requirements for the **profile page** under the section user requirements as part of functional requirements.
A separate issue is created for this purpose. You can view it in issue #19 .
Now, this issue depends on that sub-issue to be reviewed and closed.

**Comment by bahricanyesil** [comment]
We can collect all requirements under [this main requirements page](https://github.com/bounswe/bounswe2022group2/wiki/Requirements) after the review process is completed. Its structure can be enhanced and open to your contribution.

**Comment by ezgy** [comment]
I reviewed the Authentication section @bahricanyesil. I need a clarification:  
- Under the "1.1.1.5. Forgot Password" section, will the verification code sent to the user's email allow login or prompt for password reset? If it is only for successful login, will the code provided replace the user's old password?

**Comment by ezgy** [comment]
I reviewed the Profile Page section @xltvy. I have the following questions & suggestions:
- Since there are many items under each item, it might be useful to enumerate these items as well (i.e. 1.1.2.1.1 for users shall have a profile page) for easy referral to items. 
-  Under  "Interests and Knowledge" - I'm having trouble understanding the difference/connection between items 2 and 4 can you explain?
- Are _badges_ referred in "Interests and Knowledge" the same thing or different from _achievements_ referred to in "1.1.2.1.3. Achievements" ? If they are different, what does each represent? If they are the same we should come up with a single term for it.
- Will there be an option to comment and rate anonymously since students may refrain from negative comments when their names are displayed?

**Comment by xltvy** [comment]
> I reviewed the Profile Page section @xltvy. I have the following questions & suggestions:
> 
> * Since there are many items under each item, it might be useful to enumerate these items as well (i.e. 1.1.2.1.1 for users shall have a profile page) for easy referral to items.
> * Under  "Interests and Knowledge" - I'm having trouble understanding the difference/connection between items 2 and 4 can you explain?
> * Are _badges_ referred in "Interests and Knowledge" the same thing or different from _achievements_ referred to in "1.1.2.1.3. Achievements" ? If they are different, what does each represent? If they are the same we should come up with a single term for it.
> * Will there be an option to comment and rate anonymously since students may refrain from negative comments when their names are displayed?

Thanks for your detailed review. Responding for each item:
* It makes lot of sense, I have edited the [Requirements page](https://github.com/bounswe/bounswe2022group2/wiki/Requirements) of our Wiki, adding another level of enumeration for the items you have mentioned.
* "1.1.2.1.9.2 Knowledge base of the student users based on the courses taken and passed successfully and the scores from those courses shall be displayed in the profile pages." item is referring for the general definition of the knowledge base section of the profile page of a student user. On the other hand "1.1.2.1.9.4 Each completed course shall be displayed under the section of its relevant field of study in the knowledge base section of their profile page." item is referring how completed courses are displayed in the knowledge base: knowledge base is divided into parts of fields of study and each completed course is listed under its relevant field of study. If it is somehow unclear, I can edit the requirement definitions further.
* I thought badges are course specific trophies and achievements are application specific trophies. For example: When you complete five courses successfully from Maths field of study, you gain an achievement. When you are taking a Derivative course and pass a quiz with full score, you earn a badge, which is set by the educator when preparing the course material. Since the definition and basic features of both achievements and badges are not user->profile page related requirements, I have not explained them in the requirements section. The use case I have mentioned here was what I remember from our meeting, if I am not mistaken.
* Again, commenting and rating the courses and educators are done after completing the course and thus, they are course related requirements, not profile page related. So, that is why I have not added such requirements here.

**Comment by ezgy** [comment]
> > I reviewed the Profile Page section @xltvy. I have the following questions & suggestions:
> > 
> > * Since there are many items under each item, it might be useful to enumerate these items as well (i.e. 1.1.2.1.1 for users shall have a profile page) for easy referral to items.
> > * Under  "Interests and Knowledge" - I'm having trouble understanding the difference/connection between items 2 and 4 can you explain?
> > * Are _badges_ referred in "Interests and Knowledge" the same thing or different from _achievements_ referred to in "1.1.2.1.3. Achievements" ? If they are different, what does each represent? If they are the same we should come up with a single term for it.
> > * Will there be an option to comment and rate anonymously since students may refrain from negative comments when their names are displayed?
> 
> Thanks for your detailed review. Responding for each item:
> 
> * It makes lot of sense, I have edited the [Requirements page](https://github.com/bounswe/bounswe2022group2/wiki/Requirements) of our Wiki, adding another level of enumeration for the items you have mentioned.
> * "1.1.2.1.9.2 Knowledge base of the student users based on the courses taken and passed successfully and the scores from those courses shall be displayed in the profile pages." item is referring for the general definition of the knowledge base section of the profile page of a student user. On the other hand "1.1.2.1.9.4 Each completed course shall be displayed under the section of its relevant field of study in the knowledge base section of their profile page." item is referring how completed courses are displayed in the knowledge base: knowledge base is divided into parts of fields of study and each completed course is listed under its relevant field of study. If it is somehow unclear, I can edit the requirement definitions further.
> * I thought badges are course specific trophies and achievements are application specific trophies. For example: When you complete five courses successfully from Maths field of study, you gain an achievement. When you are taking a Derivative course and pass a quiz with full score, you earn a badge, which is set by the educator when preparing the course material. Since the definition and basic features of both achievements and badges are not user->profile page related requirements, I have not explained them in the requirements section. The use case I have mentioned here was what I remember from our meeting, if I am not mistaken.
> * Again, commenting and rating the courses and educators are done after completing the course and thus, they are course related requirements, not profile page related. So, that is why I have not added such requirements here.

thank you for the clarifications, I understand now. I think these will be more clear with the addition of glossary and there is no need to change anything. Great and very detailed work!

**Comment by bahricanyesil** [comment]
Waiting for @mbatuhancelik and @ahmet633 to finish the review process and close the issue.

**Comment by bahricanyesil** [comment]
> I reviewed the Authentication section @bahricanyesil. I need a clarification:
> 
> * Under the "1.1.1.5. Forgot Password" section, will the verification code sent to the user's email allow login or prompt for password reset? If it is only for successful login, will the code provide replace the user's old password?

Thank you for the review @ezgy. You are right, there is an ambiguity in that part. What I thought while writing was as follows:
1. User writes his/here mail to receive a verification code
2. User writes the verification code s/he received on a new page
3. If the verification code is matched with the sent one, the user is directed to a new page to reset the old password
4. S/he enters the new password and the old password is replaced with the new one.

If this flow is sensible and OK for you, I can revise my sentences accordingly.

**Comment by mbatuhancelik** [comment]
I proposed a prototype under #23

**Comment by ahmet633** [comment]
Currently, I don't see anything new to add. However, in the upcoming weeks there will be new things to add, old ones to edit for sure. I am more than willing to work on those improvements when they are needed.

**Comment by ezgy** [comment]
> > I reviewed the Authentication section @bahricanyesil. I need a clarification:
> > 
> > * Under the "1.1.1.5. Forgot Password" section, will the verification code sent to the user's email allow login or prompt for password reset? If it is only for successful login, will the code provide replace the user's old password?
> 
> Thank you for the review @ezgy. You are right, there is an ambiguity in that part. What I thought while writing was as follows:
> 
> 1. User writes his/her email to receive a verification code
> 2. User writes the verification code s/he received on a new page
> 3. If the verification code is matched with the sent one, the user is directed to a new page to reset the old password
> 4. S/he enters the new password and the old password is replaced with the new one.
> 
> If this flow is sensible and OK for you, I can revise my sentences accordingly.

Seems good to me, thank you for the clarification.

**Comment by bahricanyesil** [comment]
The initial user requirements are determined. Review is completed by @ezgy and corresponding responses are added as comments to the review. Now, I'm closing the issue, thanks to those who have contributed.
